pkg/networkservice/common/policyroute/server.go

```diff
@@ -0,0 +1,119 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policyroute
+
+import (
+	"context"
+
+	"sync"
+	"sync/atomic"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"gopkg.in/yaml.v2"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/fs"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+type policyrouteServer struct {
+	ctx        context.Context
+	policies   atomic.Value
+	once       sync.Once
+	configPath string
+}
+
+func (p *policyrouteServer) init(ctx context.Context) {
+	p.policies.Store([]*networkservice.PolicyRoute{})
+	updatePrefixes := func(bytes []byte) {
+		if bytes == nil {
+			p.policies.Store([]*networkservice.PolicyRoute{})
+		}
+		var source []*networkservice.PolicyRoute
+		err := yaml.Unmarshal(bytes, &source)
+		if err != nil {
+			log.FromContext(ctx).Errorf("Cannot unmarshal policies, err: %v", err.Error())
+			return
+		}
+		p.policies.Store(source)
+	}
+	updateCh := fs.WatchFile(p.ctx, p.configPath)
+	updatePrefixes(<-updateCh)
+	go func() {
+		for update := range updateCh {
+			updatePrefixes(update)
+		}
+	}()
+}
+
+// NewServer creates a NetworkServiceServer that will put the routing policies to connection context
+func NewServer(ctx context.Context, setters ...ServerOption) networkservice.NetworkServiceServer {
+	server := &policyrouteServer{
+		configPath: "/etc/route-policies/config.yaml",
+		ctx:        ctx,
+	}
+	for _, setter := range setters {
+		setter(server)
+	}
+
+	return server
+}
+
+func (p *policyrouteServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	p.once.Do(func() {
+		p.init(ctx)
+	})
+
+	conn := request.GetConnection()
+	if conn.GetContext() == nil {
+		conn.Context = &networkservice.ConnectionContext{}
+	}
+	if conn.GetContext().GetIpContext() == nil {
+		conn.GetContext().IpContext = &networkservice.IPContext{}
+	}
+	ipContext := conn.GetContext().GetIpContext()
+
+	// Update policies
+	// Remove old IP addresses
+	policies := p.policies.Load().([]*networkservice.PolicyRoute)
+	for _, p := range ipContext.Policies {
+		if p.From != "" {
+			for s := range ipContext.SrcIpAddrs {
+				if ipContext.SrcIpAddrs[s] == p.From {
+					ipContext.SrcIpAddrs = append(ipContext.SrcIpAddrs[:s], ipContext.SrcIpAddrs[s+1:]...)
```

Comment:
Could you explain why do we need to delete the src ip address in this if branch?

Reply:
So, we assume that policies can be dynamic. If we on the refresh cycle, we need to update policies and SrcIPAddreses according to new values.  On the previous `Request` we already set SrcIpAddress - https://github.com/networkservicemesh/sdk/pull/1216/files#diff-7db704fbded091b2c2a44d017b420651c29c4c949321bd5903825fc80fde77a2R116
And on `refresh` we need to delete previous IP addresses from the context and then add new ones. Because we have no idea what policy values ​​we have now (they are dynamic).